### PR TITLE
[GLUTEN-2031][VL] Support rand function

### PR DESCRIPTION
**Pull request by PHILO-HE**
This PR proposes the support for rand function, only considering the usage with no user-specified seed, i.e., `rand()`.

There will be a separate PR to support `rand(seed)`. It requires combing user-specified seed with partition index to serve as the final seed for random generator in native lib. Thus, we can achieve global cross-partitions random distribution. If only user-specified seed is used, same random sequence will be generated for each partition.
See [spark source code](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala#L45).

## How was this patch tested?
Spark UTs.

### CODE REVIEW

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala [hunk 1]
```diff
@@ -223,48 +223,32 @@ object BackendSettings extends BackendSettingsApi {
...
-   * Check whether plan is Count(1).
+   * Check whether a plan needs to be offloaded even though they have empty input schema, e.g,
+   * Sum(1), Count(1), rand(), etc.
    * @param plan:
    *   The Spark plan to check.
-   * @return
-   *   Whether plan is an Aggregation of Count(1).
    */
-  private def isCount1(plan: SparkPlan): Boolean = {
+  private def mayNeedOffload(plan: SparkPlan): Boolean = {
```
Comment (winningsix) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:231]: Minor: ```isOffloadNeededWithoutSchema``` better?

**Review by winningsix — COMMENTED** [review]
Do we need some tests for ```rand``` function?

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala [hunk 2]
```diff
@@ -223,48 +223,32 @@ object BackendSettings extends BackendSettingsApi {
...
+            }
+          })
+      case p: ProjectExec if p.projectList.nonEmpty =>
+        p.projectList.forall(
+          expr =>
+            expr match {
+              // Block directly falling back the below functions by FallbackEmptySchemaRelation.
+              case _: Rand => true
+              case _ => false
+            })
```
Comment (zhouyuan) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:251]: these changes seems not related with `rand()` ?

Comment (PHILO-HE) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:251]: As `rand()` has no child output, we need this to avoid make it fall back by `fallbackOnEmptySchema` rule.

**Comment by PHILO-HE** [comment]
Run Gluten Clickhouse CI

File: backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCHParquetSuite.scala [hunk 3]
```diff
@@ -376,8 +376,8 @@ class GlutenClickHouseTPCHParquetSuite extends GlutenClickHouseTPCHAbstractSuite
   }
 
   test("test 'function rand'") {
-    runSql("select rand(), rand(1), rand(null) from range(10)")(
-      checkOperatorMatch[ProjectExecTransformer])
+    // TODO: add rand(1), rand(null) test back.
+    runSql("select rand() from range(10)")(checkOperatorMatch[ProjectExecTransformer])
```
Comment (PHILO-HE) [inline backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCHParquetSuite.scala:380]: Hi @taiyang-li, this PR removes CH's testing for `rand(seed)`. This PR make such case fall back. See PR description. Please let me know if you have other ideas. Thanks!

Comment (PHILO-HE) [reply backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCHParquetSuite.scala:380]: cc @zzcclp

Comment (taiyang-li) [reply backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCHParquetSuite.scala:380]: @PHILO-HE Maybe we can add a if in RandTransformer for clickhouse backend to make sure CH do not fallback.

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCHParquetSuite.scala (hunk 3 above)
Comment (PHILO-HE) [reply backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCHParquetSuite.scala:380]: > @PHILO-HE Maybe we can add a if in RandTransformer for clickhouse backend to make sure CH do not fallback.

Hi @taiyang-li, does CH already handle the issue as mentioned in description? I'm assuming CH has the same issue which needs a separate fix, so I did this change.

Comment (taiyang-li) [reply backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCHParquetSuite.scala:380]: Thanks for reminding. Currently CH doesn't need the feature you mentioned in PR. Because CH rand function doesn't support user speicified seed.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala [hunk 4]
```diff
@@ -233,48 +233,35 @@ object BackendSettings extends BackendSettingsApi {
...
-              expression => {
-                expression.aggregateFunction match {
-                  case c: Count => c.children.size == 1 && c.children.head.equals(Literal(1))
-                  case _ => false
-                }
-              }) =>
-        true
-      case _ =>
-        false
+  private def mayNeedOffload(plan: SparkPlan): Boolean = {
```
Comment (rui-mo) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:241]: About the naming, is there any case where this function returns true but cannot be offloaded? Why **may**NeedOffload?

Comment (rui-mo) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:241]: OK, I see. Current name is good as whether to offload is not decided.

**Review by rui-mo — APPROVED** [review]
